### PR TITLE
Use Docker multistage build

### PR DESCRIPTION
**Pull request by jcberthon**
I've found your project pretty cool and I wanted to contribute to it.

I know my ways with Docker and therefore I'm proposing small optimisations.

What I've done is transform the Dockerfile into a multistage build. Meaning the first stage is only there to build the application and is "discarded" (1), the second stage picks up just the build data, nothing else.

Note that I am not at all experienced with npm, node or purescript/pulp. So one can do a better job at packaging a production application. I've grabbed the entire `output/` and `node_modules/` directories into the final container image and I'm pretty sure this is too much.

The total build time is about the same, but the multistage approach provides the following advantages:

* Final image (the "production image") **size** is **reduced from 830MB to 120MB** (both uncompressed)
* Container **startup** time went **down from about 11s to 4s** (on my underpower laptop with a SSD)
* Container resource usage has been improved:
  * **RAM** usage went **down from 14MB to 2MB**
  * **IO** usage for startup went **down from 31.9MB to 3.8MB**

In addition by running with nginx instead of http-server, this is using a more robust, battle-proven web server with a good security track. On top of that, the nginx server is running as a standard user and not as root inside the container. So all the better.

Measurements done on Fedora 28, using btrfs storage backend for Docker CE 18.06.0, on a laptop with a SSD and dual core mobile CPU.

*(1): The first stage is not really discarded, it is still cached by Docker so it can still accelerate subsequent builds, but only a subset of the data from 1st stage will be use in 2nd stage.*

### CODE REVIEW

**Comment by jcberthon** [comment]
Note: I have tested it that it has the same behaviour when running as the containers generated with the previous Dockerfile. But as with the original Dockerfile and as with my changes, when I run it against a private instance of GitLab, I always get an empty list.

image

So both the original and new variant of  the container do not really work with my GitLab instance, but at least my changes are not making it worst ;-)

File: Dockerfile [hunk 1]
```diff
@@ -1,11 +1,18 @@
-FROM node:10.6.0
+# Stage 1
+FROM node:10.8 as yarnbuild
 
 WORKDIR /data
 COPY . /data/
 
 RUN npm install
 RUN npm run install
```
Comment (f-f) [inline Dockerfile]: I think both these two lines can be replaced with a single `yarn install` command, see the one that we're using [here](https://github.com/KSF-Media/gitlab-dashboard/blob/master/.gitlab-ci.yml#L18)

**Review by f-f — APPROVED** [review]
Thanks @jcberthon!

I left a small comment, but otherwise looks good to me 👍
Thanks for the profiling as well 👏

**Comment by f-f** [comment]
@jcberthon 

**About the empty page**: are you providing to the page the correct url-params as listed [here](https://github.com/KSF-Media/gitlab-dashboard#demo)?
If yes, do you get any error message in the javascript console?

**About having a better optimized build**: you should be able to avoid copying the `output` folder, as the PureScript production build gets outputted to `app.js`: see [this script](https://github.com/KSF-Media/gitlab-dashboard/blob/master/scripts/install) that gets run on `yarn install`.  
However, we are aware that copying the `node_modules` is really wasteful, and that could be solved by packaging the whole npm build, e.g. with [Parcel](https://parceljs.org/) (but in that case we wouldn't be able to optimize the Purescript build, so it's not super simple). 
We have an internal project which uses Purescript + Parcel, so we soon might switch this one to that as well.

**Comment by aniyazov93** [comment]
I sometimes get blank pages too. In js console attempts to fetch tasks from project and error:

`Polling failed with error: Failed to parse jobs: (NonEmptyList (NonEmpty (ErrorAtProperty "pipeline" (ErrorAtProperty "status" (ForeignError "Expected \"running\" tag for unary constructor literal running"))) ((ErrorAtProperty "pipeline" (ErrorAtProperty "status" (ForeignError "Expected \"pending\" tag for unary constructor literal pending"))) : (ErrorAtProperty "pipeline" (ErrorAtProperty "status" (ForeignError "Expected \"success\" tag for unary constructor literal success"))) : (ErrorAtProperty "pipeline" (ErrorAtProperty "status" (ForeignError "Expected \"failed\" tag for unary constructor literal failed"))) : (ErrorAtProperty "pipeline" (ErrorAtProperty "status" (ForeignError "Expected \"canceled\" tag for unary constructor literal canceled"))) : (ErrorAtProperty "pipeline" (ErrorAtProperty "status" (ForeignError "Expected \"skipped\" tag for unary constructor literal skipped"))) : Nil)))`

I think in some jobs presents unexpected values and script hangs, but after some time bad task goes away (script poll only 100 items) and dashboard begin work again.

**Comment by f-f** [comment]
@radium88 thanks a lot, this is very helpful! What Gitlab version are you running on?

**Comment by aniyazov93** [comment]
gitlab-ce omnibus 11.1.4 (63daf37)

**Comment by f-f** [comment]
@radium88 cool, thanks. From the stack trace it looks like the "status" of the "pipeline" for a given Job (see example response [here](https://docs.gitlab.com/ee/api/jobs.html)) is not in the set of possible "Pipeline status" (see e.g. [here](https://docs.gitlab.com/ee/api/pipelines.html) for a list), so this might be a documentation hole on Gitlab's part.

I wonder how we could replicate this? 🤔 

@jcberthon do you see the same error?

**Comment by aniyazov93** [comment]
The status of pipelines, one of: running, pending, success, failed, canceled, skipped. 
This list missing 'manual' type of task, which I'm using :)

**Comment by aniyazov93** [comment]
And I think it can be easily reproduced by creating task with options
  when: manual
  allow_failure: false

**Comment by aniyazov93** [comment]
I think we've get a little off-topic of this PR, so I created new issue :)

File: Dockerfile (hunk 1 above)
Comment (jcberthon) [reply Dockerfile]: Great, I will try it out and push a new change then. Thanks.

**Comment by jcberthon** [comment]
Regarding the blank page problem, I will check my JS console for any errors and either comment on @radium88 issue or create a new one if it is a different error.

**Comment by jcberthon** [comment]
@f-f I see another error and it seems that maybe Firefox is blocking the request. I will open a new issue for that, see #32.

File: Dockerfile (hunk 1 above)
Comment (jcberthon) [reply Dockerfile]: it works indeed fine. So I've pushed a new commit with this change.

**Comment by jcberthon** [comment]
Hi @f-f 

I have updated the PR with your 2 recommendations (using yarn for the build and not copying the output directory). The image is now even a bit smaller, only 100MB uncompressed.

If you need any other help with Docker or containers, feel free to ask :-)

**Comment by f-f** [comment]
Thanks a lot @jcberthon! :)

**Pull request merged**